system_zero/src/alu/subtraction.rs

```diff
@@ -10,15 +10,45 @@ use crate::registers::alu::*;
 use crate::registers::NUM_COLUMNS;
 
 pub(crate) fn generate_subtraction<F: PrimeField64>(values: &mut [F; NUM_COLUMNS]) {
-    // TODO
+    let in_1 = values[COL_SUB_INPUT_0].to_canonical_u64() as u32;
+    let in_2 = values[COL_SUB_INPUT_1].to_canonical_u64() as u32;
+
+    // in_1 - in_2 == diff - br*2^32
+    let (diff, br) = in_1.overflowing_sub(in_2);
+
+    let diff_1 = F::from_canonical_u16(diff as u16);
+    let diff_2 = F::from_canonical_u16((diff >> 16) as u16);
+
+    values[COL_SUB_OUTPUT_0] = F::from_canonical_u16(diff as u16);
+    values[COL_SUB_OUTPUT_1] = F::from_canonical_u16((diff >> 16) as u16);
+    values[COL_SUB_OUTPUT_2] = F::from_canonical_u16(br as u16);
 }
 
 pub(crate) fn eval_subtraction<F: Field, P: PackedField<Scalar = F>>(
     local_values: &[P; NUM_COLUMNS],
     yield_constr: &mut ConstraintConsumer<P>,
 ) {
     let is_sub = local_values[IS_SUB];
-    // TODO
+    let in_1 = local_values[COL_SUB_INPUT_0];
+    let in_2 = local_values[COL_SUB_INPUT_1];
+    let out_1 = local_values[COL_SUB_OUTPUT_0];
+    let out_2 = local_values[COL_SUB_OUTPUT_1];
+    let out_br = local_values[COL_SUB_OUTPUT_2];
+
+    let base = F::from_canonical_u64(1 << 16);
+    let base_sqr = F::from_canonical_u64(1 << 32);
+    // Note that this can't overflow. Since each output limb has been
+    // range checked as 16-bits
+    let out = (out_br * base_sqr + out_1) - out_2 * base;
+
+    // NB: Not clear how to compute in_1 - in_2 in one expression for
+    // PackedFields: If in_1 < in_2 then the sign extension will
+    // happen inside the big field, which is probably wrong. Instead,
+    // we *first* subtract in_2 from the expected output, and then
+    // subtract in_1; the result then should be zero.
+    yield_constr.constraint(is_sub * ((out + in_2) - in_1));
+
+    yield_constr.constraint(is_sub * out_br * (P::ONES - out_br));
```

Comment:
I think we can omit this, since the borrow output register is an alias for a `col_bit` register, and those will be constrained to be 0 or 1 by the "boolean unit".

(We haven't actually added those constraints yet, but that's the plan at least.)

Reply:
Yeah I think so 👍